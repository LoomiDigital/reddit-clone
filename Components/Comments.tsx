import React from "react";
import { useComments } from "@d20/hooks/useComments";
import { PostAttributesFragment } from "@d20/generated/graphql";

import CommentCard from "./CommentCard";
import { CommentLoader } from "./Loaders";

interface Props {
  post: PostAttributesFragment;
}

const Comments = ({ post }: Props) => {
  const { submitComment, loading, comments, session, isValid, register } =
    useComments(post);

  return (
    <>
      <div className="mt-5 rounded-t-md border border-b-0 border-t-0 border-gray-300 bg-white p-5 pl-16">
        <p className="text-sm">
          Comments as{" "}
          <span className="text-red-500">{session?.user?.name}</span>
        </p>
        <form onSubmit={submitComment} className="flex flex-col space-y-2">
          <textarea
            disabled={!session}
            {...register("comment", { required: true })}
            className="h-24 rounded-sm border border-gray-200 p-2 pl-4 outline-none focus-within:border-gray-900 disabled:bg-gray-50"
            placeholder={
              session ? "What are your thoughts?" : "Please log in to comment"
            }
          ></textarea>
          <button
            disabled={!isValid || !session}
            type="submit"
            name="comment"
            className="rounded-full bg-red-500 p-3 font-semibold text-white disabled:cursor-default disabled:bg-gray-500 disabled:text-gray-300"
          >
            Comment
          </button>
        </form>
      </div>
      <div className="rounded-b-md border border-t-0 border-gray-300 bg-white px-10 py-5">
        <hr className="py-2" />
        {loading ? (
          <CommentLoader length={10} />
        ) : (
          comments?.map((comment) => (
            <CommentCard key={comment?.id} comment={comment!} />
          ))
        )}
      </div>
    </>
  );
};

export default Comments;
