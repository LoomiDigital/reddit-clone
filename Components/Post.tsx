import React from "react";

interface Props {
  post: Post;
}

function Post({ post }: Props) {
  return <div>Post</div>;
}

export default Post;
