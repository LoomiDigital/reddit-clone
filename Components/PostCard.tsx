import React, { SyntheticEvent, useEffect, useState } from "react";
import Link from "next/link";

import { loadPage } from "@d20/utils/loadPage";
import { useAddVote } from "@d20/hooks/useAddVote";
import { useComments } from "@d20/hooks/useComments";
import { PostAttributesFragment } from "@d20/generated/graphql";

import TimeAgo from "react-timeago";

import {
  ArrowDownIcon,
  ArrowUpIcon,
  BookmarkIcon,
  ChatBubbleOvalLeftEllipsisIcon,
  EllipsisHorizontalIcon,
  GiftIcon,
  ShareIcon,
} from "@heroicons/react/24/outline";
import Avatar from "./Avatar";

interface Props {
  post: PostAttributesFragment;
}

const PostCard = ({ post }: Props) => {
  const loadSubredditPage = (e: SyntheticEvent) => {
    loadPage(e, `/subreddit/${post.subreddit_topic}`);
  };

  const [hasMounted, setHasMounted] = useState<boolean>(false);
  const { upVote, vote, displayVotes } = useAddVote(post);
  const { comments, loading } = useComments(post);

  useEffect(() => {
    setHasMounted(true);
  }, []);

  return (
    <Link passHref href={`/post/${post.id}`}>
      <div className="flex cursor-pointer rounded-md border border-gray-300 bg-white shadow-sm hover:border hover:border-gray-600">
        <div className="flex flex-col items-center justify-start space-y-1 rounded-l-md bg-gray-50 p-4 text-gray-400">
          <ArrowUpIcon
            onClick={(e) => upVote(e, true)}
            title="upvote"
            className={`voteButtons cursor-pointer hover:text-red-400 ${
              vote && "text-red-400"
            }`}
          />

          <p title="votes" className="votes text-xs font-bold text-black">
            {displayVotes}
          </p>

          <ArrowDownIcon
            onClick={(e) => upVote(e, false)}
            title="downvote"
            className={`voteButtons cursor-pointer hover:text-blue-400 ${
              !vote && "text-blue-400"
            }`}
          />
        </div>
        <div className="p-3 pb-1">
          <div className="flex items-center space-x-2">
            <Avatar seed={post?.subreddit_topic!} />
            <p className="text-xs text-gray-400">
              <span
                onClick={(e) => loadSubredditPage(e)}
                className="font-bold text-black hover:text-blue-400 hover:underline"
              >
                r/{post.subreddit_topic}
              </span>

              <span>
                â€¢ Posted by u/{post.username}{" "}
                {hasMounted ? (
                  <TimeAgo date={post.created_at} autoPlay={false} />
                ) : (
                  "Loading..."
                )}
              </span>
            </p>
          </div>

          <div className="py-4">
            <h2 className="text-xl font-semibold">{post.title}</h2>
            <p className="mt-2 text-sm font-light">{post.body}</p>
          </div>
          {post.image && <img className="w-full" src={post.image} alt="" />}

          <div className="flex space-x-4 text-gray-400 ">
            <div className="postButtons">
              <ChatBubbleOvalLeftEllipsisIcon className="h-6 w-6" />
              <p>{loading ? 0 : comments?.length} Comments</p>
            </div>
            <div className="postButtons">
              <GiftIcon className="h-6 w-6" />
              <p className="hidden sm:inline">Award</p>
            </div>
            <div className="postButtons">
              <ShareIcon className="h-6 w-6" />
              <p className="hidden sm:inline">Share</p>
            </div>
            <div className="postButtons">
              <BookmarkIcon className="h-6 w-6" />
              <p className="hidden sm:inline">Save</p>
            </div>
            <div className="postButtons">
              <EllipsisHorizontalIcon className="h-6 w-6" />
            </div>
          </div>
        </div>
      </div>
    </Link>
  );
};

export default PostCard;
