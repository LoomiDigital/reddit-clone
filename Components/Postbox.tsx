import React, { useState } from "react";
import { useSession } from "next-auth/react";
import { useForm } from "react-hook-form";
import { toast } from "react-hot-toast";
import { LinkIcon, PhotoIcon } from "@heroicons/react/24/outline";
import {
  PostAttributesFragmentDoc,
  useAddPostMutation,
  useAddSubredditMutation,
  useAddVoteMutation,
  useGetSubredditByTopicLazyQuery,
} from "@d20/generated/graphql";

import Avatar from "./Avatar";

type Props = {
  subreddit?: string;
};

type FormData = {
  postTitle: string;
  postBody: string;
  postImage: string;
  subreddit: string;
};

function Postbox({ subreddit }: Props) {
  const { data: session } = useSession();
  const [addPost] = useAddPostMutation();
  const [addVote] = useAddVoteMutation();
  const [addSubreddit] = useAddSubredditMutation();
  const [getSubReddit] = useGetSubredditByTopicLazyQuery();

  const {
    register,
    handleSubmit,
    setValue,
    watch,
    formState: { errors },
  } = useForm<FormData>();
  const [imageBoxOpen, setImageBoxOpen] = useState<boolean>(false);

  const submitPost = handleSubmit(async (formData) => {
    const notification = toast.loading("Creating post...");

    try {
      const { data } = await getSubReddit({
        variables: {
          topic: subreddit || formData.subreddit,
        },
        fetchPolicy: "no-cache",
      });

      const subredditExists = data?.getSubredditByTopic;

      const postFields = {
        title: formData.postTitle!,
        body: formData.postBody || "",
        image: formData.postImage || "",
        username: session?.user.name!,
      };

      let newSubreddit = null;

      if (!subredditExists) {
        const { data: addSubredditData } = await addSubreddit({
          variables: {
            topic: formData.subreddit.toLowerCase(),
          },
        });
        newSubreddit = addSubredditData?.insertSubreddit!;
      }

      addPost({
        variables: {
          ...postFields,
          subreddit_id: newSubreddit?.id || subredditExists?.id!,
          subreddit_topic: newSubreddit?.topic || subredditExists?.topic!,
        },
        optimisticResponse(vars) {
          return {
            insertPost: {
              __typename: "Post",
              id: -1,
              title: vars.title,
              body: vars.body,
              image: vars.image,
              username: vars.username,
              votes: [],
              subreddit_id: -1,
              subreddit_topic: vars.subreddit_topic,
              created_at: new Date().toISOString(),
            },
          };
        },
        update: async (cache, { data: addPostData }) => {
          const fieldName = subreddit ? "postsByTopic" : "posts";

          cache.modify({
            fields: {
              [fieldName](existingPosts = {}) {
                const newPostRef = cache.writeFragment({
                  data: {
                    ...addPostData?.insertPost!,
                    votes: [
                      {
                        __typename: "Vote",
                        id: 666,
                        username: session?.user.name!,
                        upvote: true,
                      },
                    ],
                  },
                  fragment: PostAttributesFragmentDoc,
                  fragmentName: "postAttributes",
                });

                return {
                  ...existingPosts,
                  edges: [
                    {
                      __typename: "PostEdge",
                      node: newPostRef,
                      cursor: addPostData?.insertPost?.id!,
                    },
                    ...existingPosts.edges,
                  ],
                };
              },
            },
          });
        },
        onCompleted: (data) => {
          addVote({
            variables: {
              post_id: data.insertPost?.id!,
              username: session?.user.name!,
              upvote: true,
            },
          });
        },
      });

      setValue("postTitle", "");
      setValue("postBody", "");
      setValue("postImage", "");
      setValue("subreddit", "");

      toast.success("Post created!", { id: notification });
    } catch (error) {
      console.log("error", error);
      toast.error("Error creating post!", { id: notification });
    }
  });

  return (
    <form
      className="sticky top-16 z-50 rounded-md border border-gray-300 bg-white p-2"
      onSubmit={submitPost}
    >
      <div className="flex items-center space-x-3">
        <Avatar />

        <input
          {...register("postTitle", { required: true })}
          disabled={!session}
          className="w-full flex-1 rounded-md bg-gray-50 p-2 pl-5 outline-none"
          type="text"
          placeholder={
            session
              ? subreddit
                ? `Create a post in r/${subreddit}`
                : "Create a post by entering a title!"
              : "Sign in to create a post"
          }
        />
        <PhotoIcon
          onClick={() => {
            setImageBoxOpen(!imageBoxOpen);
          }}
          className={`h-6 cursor-pointer text-gray-300 ${
            imageBoxOpen && "text-blue-500"
          }`}
        />
        <LinkIcon className="h-6 text-gray-300" />
      </div>
      {!!watch("postTitle") && (
        <div className="flex flex-col py-2">
          <div className="flex items-center px-2">
            <p className="min-w-[90px]">Body:</p>
            <input
              className="m-2 flex-1 bg-blue-50 p-2 outline-none"
              {...register("postBody")}
              type="text"
              placeholder="Text (optional)"
            />
          </div>
          {!subreddit && (
            <div className="flex flex-col py-2">
              <div className="flex items-center px-2">
                <p className="min-w-[90px]">Subreddit:</p>
                <input
                  className="m-2 flex-1 bg-blue-50 p-2 outline-none"
                  {...register("subreddit", { required: true })}
                  type="text"
                  placeholder="i.e. r/nextjs"
                />
              </div>
            </div>
          )}
          {imageBoxOpen && (
            <div className="flex flex-col py-2">
              <div className="flex items-center px-2">
                <p className="min-w-[90px]">Image URL:</p>
                <input
                  className="m-2 flex-1 bg-blue-50 p-2 outline-none"
                  {...register("postImage")}
                  type="text"
                  placeholder="Optional..."
                />
              </div>
            </div>
          )}
          {Object.keys(errors).length > 0 && (
            <div className="space-y-2 p-2 text-red-500">
              {errors.postTitle?.type === "required" && (
                <p>- Post title is required</p>
              )}
              {errors.subreddit?.type === "required" && (
                <p>- A subreddit is required</p>
              )}
            </div>
          )}
          {!!watch("postTitle") && (
            <button
              type="submit"
              className="w-full rounded-full bg-blue-500 p-2 text-white"
            >
              Create Post
            </button>
          )}
        </div>
      )}
    </form>
  );
}

export default Postbox;
