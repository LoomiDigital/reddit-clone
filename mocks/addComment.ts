import { MockedResponse } from "@apollo/client/testing";
import { AddCommentDocument } from "@d20/generated/graphql";

export const mockAddComment: MockedResponse = {
  request: {
    query: AddCommentDocument,
    variables: {
      post_id: 1,
      text: "This is a new comment",
      username: "Buck",
    },
  },
  result: {
    data: {
      addComment: {
        id: 3,
        text: "This is a new comment",
        post_id: 1,
        created_at: "2023-05-08T18:10:10.831966Z",
        username: "Buck",
      },
    },
  },
};
