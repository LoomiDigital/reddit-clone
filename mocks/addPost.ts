import { MockedResponse } from "@apollo/client/testing";
import { AddPostDocument } from "@d20/generated/graphql";

export const mockAddPost: MockedResponse = {
  request: {
    query: AddPostDocument,
    variables: {
      title: "This is a new post",
      body: "A brand new post!",
      image: "https://via.placeholder.com/150",
      username: "Buck",
      subreddit_id: 1,
      subreddit_topic: "testsubreddit",
    },
  },
  result: {
    data: {
      insertPost: {
        __typename: "Post",
        id: 1,
        title: "This is a new post",
        body: "A brand new post!",
        image: "https://via.placeholder.com/150",
        username: "Buck",
        subreddit_id: 1,
        subreddit_topic: "testsubreddit",
        created_at: "2021-08-15T20:00:00.000Z",
        votes: [
          {
            __typename: "Vote",
            id: -1,
            post_id: 1,
            username: "Buck",
            upvote: true,
          },
        ],
      },
    },
  },
};
