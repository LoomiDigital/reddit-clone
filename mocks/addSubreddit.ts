import { MockedResponse } from "@apollo/client/testing";
import { AddSubredditDocument } from "@d20/generated/graphql";

export const mockAddSubreddit: MockedResponse = {
  request: {
    query: AddSubredditDocument,
    variables: {
      topic: "testsubreddit",
    },
  },
  result: {
    data: {
      insertSubreddit: {
        __typename: "Subreddit",
        id: 1,
        post_id: 1,
        topic: "testsubreddit",
        created_at: "2021-08-15T20:00:00.000Z",
      },
    },
  },
};
