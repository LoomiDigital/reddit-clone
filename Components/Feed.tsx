import React from "react";
import { useQuery } from "@apollo/client";
import { GET_POSTS, GET_POSTS_BY_TOPIC } from "@d20/graphql/queries";

import Post from "./Post";

type Props = {
  posts?: Post[];
  topic?: string;
};

function Feed({ posts, topic }: Props) {
  const query = !topic ? GET_POSTS : GET_POSTS_BY_TOPIC;
  const options = !topic
    ? {}
    : {
        variables: {
          topic,
        },
      };
  const { data } = useQuery(query, options);

  const realTimePosts: Post[] = !topic
    ? data?.getPostList
    : data?.getPostsByTopic;

  return (
    <div className="mt-5 space-y-4">
      {realTimePosts
        ? realTimePosts?.map((post) => (
            <div key={post.id}>
              <Post post={post} />
            </div>
          ))
        : posts?.map((post) => (
            <div key={post.id}>
              <Post post={post} />
            </div>
          ))}
    </div>
  );
}

export default Feed;
