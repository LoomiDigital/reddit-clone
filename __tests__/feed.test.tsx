import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { SessionProvider } from "next-auth/react";
import { mockPostConnection, mockPostsResponse } from "@d20/mocks/getPosts";

import Feed from "@d20/Components/Feed";
import PostCard from "@d20/Components/PostCard";
import { PostLoader } from "@d20/Components/Loaders";

jest.mock("@d20/Components/PostCard");
jest.mock("@d20/Components/Loaders", () => ({
  PostLoader: jest.fn(),
}));

describe("Feed component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should handle the loading state", async () => {
    render(
      <MockedProvider>
        <SessionProvider
          session={{
            expires: "2021-10-10",
            user: {
              name: "Buck",
              expires: "2021-10-10",
              email: "user@test.com",
              address: "123 Fake St",
              image: "https://via.placeholder.com/150",
            },
          }}
        >
          <Feed
            loading={true}
            loadingRef={() => {}}
            posts={mockPostConnection.edges}
          />
        </SessionProvider>
      </MockedProvider>
    );

    expect(PostLoader).toHaveBeenCalled();
  });

  it("should render the correct number of posts", async () => {
    render(
      <MockedProvider mocks={[mockPostsResponse]}>
        <SessionProvider
          session={{
            expires: "2021-10-10",
            user: {
              name: "Buck",
              expires: "2021-10-10",
              email: "user@test.com",
              address: "123 Fake St",
              image: "https://via.placeholder.com/150",
            },
          }}
        >
          <Feed
            loading={false}
            loadingRef={() => {}}
            posts={mockPostConnection.edges}
          />
        </SessionProvider>
      </MockedProvider>
    );

    expect(PostCard).toHaveBeenCalledTimes(2);
  });
});
