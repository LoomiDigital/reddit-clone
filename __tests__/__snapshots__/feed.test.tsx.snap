// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed component should render correctly 1`] = `
<div
  className="mt-5 space-y-4"
>
  <div />
  <div />
  <div
    className="flex flex-col"
  />
</div>
`;
